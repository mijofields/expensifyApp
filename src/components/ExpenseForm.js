import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


const now = moment();

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        };
    }

    handleDescriptionChange = (e) => {
        let description = e.target.value;
        this.setState(() => ({description}));
    };

    handleNoteChange = (e) => {
        let note = e.target.value;
        this.setState(() => ({note}));
    };

    handleAmountChange = (e) => {
        let amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    };

    handleDateChange = (createdAt) =>{
        if (createdAt){
            this.setState(() => ({createdAt}));
        }
    };

    handleFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}));

    };

    handleSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please set both a description and an amount'}));
        } else {
            this.setState(() => ({error: undefined}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }

    }


    render () {
        return (
            
                
                <form className="form" onSubmit={this.handleSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <input
                    type="text"
                    className="text-input"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.handleDescriptionChange}
                    />
                    <input 
                    type="text"
                    className="text-input"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.handleAmountChange}
                    />
                    <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.handleDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.handleFocusChange}
                    numberOfMonths= {1}
                    isOutsideRange={() => false}
                    />
                    <textarea
                    className="textarea"
                    placeholder="add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.handleNoteChange}
                    />
                        <div>
                            <button className="button">Save Expense</button>
                        </div>
                </form>
            
        )
    }
}