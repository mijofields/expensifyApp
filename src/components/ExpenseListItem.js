import React from 'react';
import { Link } from 'react-router-dom';


const ExpenseListItem = ({ id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>Cost: {amount}, Date: {createdAt} </p>
        <p>{id}</p>
    </div>
);


export default ExpenseListItem;
