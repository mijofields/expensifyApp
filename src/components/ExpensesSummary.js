import React from 'react';
import { connect } from 'react-redux';
import getExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';


export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';

    return (
    <div>
        <h2>Viewing {expensesCount} {expenseWord} totaling {numeral(expensesTotal).format('$0,0.00')} </h2>
    </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
};


export default connect(mapStateToProps)(ExpensesSummary);