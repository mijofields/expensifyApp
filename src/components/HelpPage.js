import React from 'react';

const HelpPage = () => (
    <div>
        This will be the <strong>help</strong> component
    </div>
); 

export default HelpPage;