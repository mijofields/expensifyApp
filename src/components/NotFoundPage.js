import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        This will be the <strong>404 Error</strong> component
        Go <Link to="/">Home?</Link>
        {/* Link does clientside rendering */}
    </div>
);

export default NotFoundPage;