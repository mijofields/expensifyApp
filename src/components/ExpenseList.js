import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//regular unconnected component
export const ExpenseList = (props) => (
    <div>
       {props.expenses.length=== 0 ? (
           <p>No Expenses</p>
       ) : (
        props.expenses.map((expense) => {
            return <ExpenseListItem {...expense} key={expense.id} />;
       })
    ) 
}
    </div>
);

//function mapping state to props

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

//connected Component
export default connect(mapStateToProps)(ExpenseList);