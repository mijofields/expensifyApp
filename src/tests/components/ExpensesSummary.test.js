import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test(`should correctly render ExpensesSummary with one expense`, () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={250} /> );
    expect(wrapper).toMatchSnapshot();
});

test(`should correctly render ExpensesSummary with multiple expense`, () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={15} expensesTotal={250054} /> );
    expect(wrapper).toMatchSnapshot();
});