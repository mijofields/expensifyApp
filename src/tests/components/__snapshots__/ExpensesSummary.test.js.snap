// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with multiple expense 1`] = `
<div>
  <h2>
    Viewing 
    15
     
    expenses
     totaling 
    $250,054.00
     
  </h2>
</div>
`;

exports[`should correctly render ExpensesSummary with one expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     
    expense
     totaling 
    $250.00
     
  </h2>
</div>
`;
